tools/src/main/java/org/apache/kafka/trogdor/coordinator/TaskManager.java

```diff
@@ -322,42 +383,13 @@ public void createTask(final String id, TaskSpec spec)
         }
 
         @Override
-        public Void call() throws Exception {
-            if (id.isEmpty()) {
-                throw new InvalidRequestException("Invalid empty ID in createTask request.");
-            }
-            ManagedTask task = tasks.get(id);
-            if (task != null) {
-                if (!task.spec.equals(spec)) {
-                    throw new RequestConflictException("Task ID " + id + " already " +
-                        "exists, and has a different spec " + task.spec);
-                }
-                log.info("Task {} already exists with spec {}", id, spec);
-                return null;
-            }
-            TaskController controller = null;
-            String failure = null;
+        public ManagedTask call() throws Exception {
             try {
-                controller = spec.newController(id);
-            } catch (Throwable t) {
-                failure = "Failed to create TaskController: " + t.getMessage();
+                TaskController controller = spec.newController(id);
+                return new ManagedTask(id, spec, controller, ManagedTaskState.PENDING);
+            } catch (Exception t) {
+                throw new BadRequestException("Failed to create TaskController: " + t.getMessage());
```

Comment:
This exception gets logged and then ignored by `createAndScheduleTasksAtomic`, right?  I think it would be better to keep the original behavior where we create a task, and mark it as done and failed.  Then, at least, the user could see the error message if something failed.  The error would appear in the failed task status.  The user shouldn't have to read the logfile to see the failure message.

Reply:
Oh, nice catch. I think we should re-raise so it gets propagated and returned as a 400 to the caller. It's also not very atomic right now, is it?

Reply:
I don't think it needs to be atomic.  What I would suggest is something like:

1. Do basic validation that no task IDs are null, the JSON is well-formed, etc.  Most of this is done by jetty automatically, but we do need to manually check that the IDs aren't null, and that there is at least one task in the array.
2. Atomically, check to make sure that every task ID that we're trying to create either:
a. doesn't already exist, or
b. does exist, and with an identical task spec to what we sent
If anything doesn't match, we throw an exception and return an HTTP error for everything.
3. Create each task.  If there is an error during task creation, the task is just created in DONE state with error set to the error.
4. Return 201 CREATED.

Reply:
Okay agreed, that makes sense, especially for step 3. If a user sends 4 tasks and two of them are configured funnily, he should be able to see what's wrong with them properly